envoy/api/v2/auth/rbac.proto

```diff
@@ -0,0 +1,180 @@
+//         paths: ["/products/*"]
+//         methods: ["GET", "HEAD"]
+//         constraints:
+//         - key: request.http.headers["version"]
+//           value: ["v1", "v2"]
+//
+// A ServiceRoleBinding specification includes two parts:
+// * "roleRef" refers to a ServiceRole object.
+// * A list of "subjects" that are assigned the roles.
+//
+// A subject can be a "user" or a "group", or represented with a set of custom "properties". The
+// names of a property used to identify a subject in ServiceRoleBinding map to attributes defined in
+// AttributeContext
+// (https://github.com/envoyproxy/data-plane-api/blob/master/envoy/service/auth/v2/attribute_context.proto)
+// as follows:
+//    user: source.principal
+//    group is currently undefined. It is reserved to be used in the future.
+//
+// Custom property name directly refers to the attribute name in AttributeContext. For example,
+// source.service, source.labels["app"], request.http.headers["version"].
+//
+// Below is an example of ServiceRoleBinding object "test-binding-products", which binds two
+// subjects to ServiceRole "product-viewer":
+//   * User (service account) "cluster.local/ns/default/sa/bookinfo-reviews"
+//   * Service "reviews.default.svc.cluster.local" at version "v1".
+//
+//     apiVersion: "config.istio.io/v1alpha1"
+//     kind: ServiceRoleBinding
+//     metadata:
+//       name: test-binding-products
+//       namespace: default
+//     spec:
+//       subjects:
+//       - user: cluster.local/ns/default/sa/bookinfo-reviews
+//       - properties:
+//           source.service: "reviews.default.svc.cluster.local"
+//           source.labels["version"]: "v1"
+//       roleRef:
+//         kind: ServiceRole
+//         name: "products-viewer"
+
+// ServiceRole specification contains a list of access rules (permissions).
+// This represent the "Spec" part of the ServiceRole object. The name and namespace
+// of the ServiceRole is specified in "metadata" section of the ServiceRole object.
+message ServiceRole {
+  // Required. The set of access rules (permissions) that the role has.
+  repeated AccessRule rules = 1;
+}
+
+// AccessRule defines a permission to access a list of services.
+message AccessRule {
+  // Required. A list of service names.
+  // Exact match, prefix match, and suffix match are supported for service names.
+  // For example, the service name "bookstore.mtv.svc.cluster.local" matches
+  // "bookstore.mtv.svc.cluster.local" (exact match), or "bookstore*" (prefix match),
+  // or "*.mtv.svc.cluster.local" (suffix match).
+  // If set to ["*"], it refers to all services in the namespace.
+  repeated string services = 1;
+
+  // Optional. A list of HTTP paths or gRPC methods.
+  // gRPC methods must be presented as fully-qualified name in the form of
+  // packageName.serviceName/methodName.
+  // Exact match, prefix match, and suffix match are supported for paths.
+  // For example, the path "/books/review" matches
```

Comment:
Extract the combination of (exact match, prefix match, suffix match) to a message with `oneof` like [this](https://github.com/envoyproxy/data-plane-api/blob/master/STYLE.md) (scroll down for oneof)? It is more self descriptive than a comment.

Reply:
Yes I understand this will change the language slightly though I think it's worth. The idea is keep this extensible and self describing. For example, someone may add regex later but you don't want always do regex.